src/algorithms/ctmrg/ctmrg.jl

```diff
@@ -183,45 +189,28 @@ There are two modes of expansion: `M = :sequential` and `M = :simultaneous`.
 The first mode expands the environment in one direction at a time, for convenience towards
 the left. The second mode expands the environment in all four directions simultaneously.
 """
-function ctmrg_expand(state, envs::CTMRGEnv{C,T}, ::SequentialCTMRG) where {C,T}
-    Qtype = tensormaptype(spacetype(C), 3, 3, storagetype(C))
-    Q_sw = Zygote.Buffer(envs.corners, Qtype, axes(state)...)
-    Q_nw = Zygote.Buffer(envs.corners, Qtype, axes(state)...)
-
-    directions = collect(Iterators.product(axes(state)...))
-    # @fwdthreads for (r, c) in directions
-    for (r, c) in directions
-        Q_sw[r, c] = enlarge_southwest_corner((r, c), envs, state)
-        Q_nw[r, c] = enlarge_northwest_corner((r, c), envs, state)
+function ctmrg_expand(state, envs::CTMRGEnv, alg::SequentialCTMRG)
+    drc_combinations = collect(Iterators.product([4, 1], axes(state)...))
+    return map(drc_combinations) do (dir, r, c)
+        enlarge_corner(Val(dir), (r, c), envs, state, alg)
     end
-
-    return copy(Q_sw), copy(Q_nw)
 end
-function ctmrg_expand(state, envs::CTMRGEnv{C,T}, ::SimultaneousCTMRG) where {C,T}
-    Qtype = tensormaptype(spacetype(C), 3, 3, storagetype(C))
-    Q = Zygote.Buffer(Array{Qtype,3}(undef, size(envs.corners)))
-    drc_combinations = collect(Iterators.product(axes(envs.corners)...))
-    @fwdthreads for (dir, r, c) in drc_combinations
-        Q[dir, r, c] = if dir == NORTHWEST
-            enlarge_northwest_corner((r, c), envs, state)
-        elseif dir == NORTHEAST
-            enlarge_northeast_corner((r, c), envs, state)
-        elseif dir == SOUTHEAST
-            enlarge_southeast_corner((r, c), envs, state)
-        elseif dir == SOUTHWEST
-            enlarge_southwest_corner((r, c), envs, state)
-        end
+function ctmrg_expand(state, envs::CTMRGEnv, alg::SimultaneousCTMRG)
+    drc_combinations = collect(Iterators.product(1:4, axes(state)...))
+    return map(drc_combinations) do (dir, r, c)
+        enlarge_corner(Val(dir), (r, c), envs, state, alg)
```

Comment:
I'm not sure this solution with `Val` is really what we want, this effectively replaces a runtime `if` check with a runtime dispatch, as the compiler cannot figure out which `enlarge_corner` to call from the types of `dir, r, c`. It might be the case that this is const-propagated, but I don't think this is really necessary, the overhead of the simple if statements should really be completely negligeable...

I am completely fine with moving the `if` clauses into `enlarge_corner`, I would just refrain from using dispatch to select the method here.
Am I missing some specific benefit of doing it this way?

Reply:
You're probably not missing anything :-) I just needed to rewrite the `ctmrg_expand` a bit to make it compatible with the `EnlargedCorner` struct and reduce code duplication. Then I somehow decided to move away from the if statement.

I'll move the if clause back inside `enlarge_corner` then!